### DISCUSSION TITLE
Overlapping between bodies get not detected

### DISCUSSION THREAD

**Discussion by DennisKleinhans** [dyn4j/dyn4j#183] [Q&A] [answered]
I want to check in my simulation for overlapping bodies. I thougth I get this but now I got into a few problems. 
In the video you can see a few bodies. For example the red circle overlapps with the yellow quaders because the Mass of the red circle is very high. But when I put the circle down on the ground (ground is MassType infinite) these two bodies don't overlapp. But why?

Another Problem you can see in the Video is the two yellow bodies are jointed via a distance joint together and overlapp, but it won't be detected as a overlapping. Why is this so? 

Here you can see how I check for overlapping between two bodies:

```
	@Override
	public void end(TimeStep step, PhysicsWorld<SimulationBody, ?> world) {
		collisionPairs = new ArrayList<CollisionPair<CollisionItem<SimulationBody, BodyFixture>>>();
		
		Iterator<?> it = world.getCollisionDataIterator();
		
		while(it.hasNext()) {
			WorldCollisionData<SimulationBody> collisionData = (WorldCollisionData<SimulationBody>) it.next();
			
			if(collisionData.isNarrowphaseCollision()) {
				CollisionPair<CollisionItem<SimulationBody, BodyFixture>> collisionPair = collisionData.getPair();
				
				SimulationBody bodyA = collisionPair.getFirst().getBody();
				SimulationBody bodyB = collisionPair.getSecond().getBody();
				
				if(bodiesToCheck.contains(bodyA) && bodiesToCheck.contains(bodyB)) {
					if(Math.abs(collisionData.getPenetration().getDepth()) > overlappingDistance) {
						collisionPairs.add(collisionPair);
					}
				}
			}
			
		}
		
		// get the shapes of the bodies to set a stroke on them
		for(CollisionPair<CollisionItem<SimulationBody, BodyFixture>> pair: collisionPairs) {
			Shape shape1 = shapesAndBodies.get(pair.getFirst().getBody());
			Shape shape2 = shapesAndBodies.get(pair.getSecond().getBody());
			
			paintStrokes(shape1);
			paintStrokes(shape2);
		}
		
	}
```

Here is how I select the bodies which should be checked for overlapping:

```
// add all bodies which should be checked for overlapping to a list and pass it to the OverlapListener 
		List<SimulationBody> bodiesToCheck = new ArrayList<>();
		Iterator<SimulationBody> bodyIterator = world.getBodyIterator();
		while(bodyIterator.hasNext()) {
			SimulationBody body = bodyIterator.next();
			if(!body.getFixtures().get(0).isSensor()) {
				bodiesToCheck.add(body);
			}
		}
		world.addStepListener(new OverlapListener(bodiesToCheck, shapesAndBodies, 0.01));
```

I hope anyone understand what my problem is and can help me.

https://user-images.githubusercontent.com/44137912/108057444-ef56bb80-7052-11eb-8ffe-521f6a11fe93.mp4

**Comment by wnbittle** [comment, marked as answer]
> Another Problem you can see in the Video is the two yellow bodies are jointed via a distance joint together and overlapp, but it won't be detected as a overlapping. Why is this so?

This one is pretty easy to answer:  All `Joint`s default to NOT allowing collision between the joined bodies.  You can turn this off on a per joint basis by using the `Joint.setCollisionAllowed(boolean)` method.  Of course, this would enable _solving_ of those collisions as well - which may not be what you want.  You can control this at the global level by overriding the `BroadphaseCollisionDataFilter` via the `World.setBroadphaseCollisionDataFilter(BroadphaseCollisionDataFilter)`.  The default is [`PhysicsBodyBroadphaseCollisionDataFilter`](https://github.com/dyn4j/dyn4j/blob/62c358e5c8fd414966baef3a8380f6da5ee9e14d/src/main/java/org/dyn4j/world/PhysicsBodyBroadphaseCollisionDataFilter.java#L67) which stops the collision detection at the broadphase.

> But when I put the circle down on the ground (ground is MassType infinite) these two bodies don't overlapp. But why?
new OverlapListener(bodiesToCheck, shapesAndBodies, 0.01)
if(Math.abs(collisionData.getPenetration().getDepth()) > overlappingDistance) {

For the other issue, have you tried changing your `overlappingDistance` to 0?

**Reply by DennisKleinhans** [reply]
> This one is pretty easy to answer: All Joints default to NOT allowing collision between the joined bodies. You can turn this off on a per joint basis by using the Joint.setCollisionAllowed(boolean) method. Of course, this would enable solving of those collisions as well - which may not be what you want. You can control this at the global level by overriding the BroadphaseCollisionDataFilter via the World.setBroadphaseCollisionDataFilter(BroadphaseCollisionDataFilter). The default is PhysicsBodyBroadphaseCollisionDataFilter which stops the collision detection at the broadphase.

Thank you!

> For the other issue, have you tried changing your overlappingDistance to 0?


If I set the `overlappingDistance` to 0 it works. But when I put the red circle on the yellow square, as in the video, the yellow square is pressed into the ground, creating an overlap. When I put the red circle on the floor, it is not pressed into the floor.
I ask myself why this is, is that because the ground has an infinite mass or does that have other reasons?

**Reply by wnbittle** [reply]
I'm not sure I understand.  So when you set `overlappingDistance` to 0 _everything_ works?  Or does the red circle on the ground still not highlight?  I would expect that if the red circle is resting on the ground that there should be some overlap, likely very small, hence my suggestion.

> I ask myself why this is, is that because the ground has an infinite mass or does that have other reasons?

The amount of overlap can be due to a lot of different factors.  In your scenario, it _appears_ to be due to the total weight of the red circle + the square but it's hard to tell from the information that you've provided.  If the questions is whether this is expected behavior, the answer is yes - as you stack more and more things on top of each other, the penetration will continue to grow until the bottom is wedged out (refer to the [`Pyramid`](https://github.com/dyn4j/dyn4j-samples/blob/master/src/main/java/org/dyn4j/samples/Pyramid.java) sample and change the `ph` (pyramid height) variable).  I'd recommend to make sure you are using the latest version of the library to ensure you aren't seeing bugs that's been resolved recently.

**Reply by DennisKleinhans** [reply]
> I'd recommend to make sure you are using the latest version of the library to ensure you aren't seeing bugs that's been resolved recently.

Yes, I am using the latest version


I'm trying to rephrase my question. In this video, one square is very heavy and the other is not. If I now put the heavy square on the lighter one, it is pressed into the ground and overlaps for a short time, but when I put the heavy square on the ground there is no overlap. 
My question is why there is an overlap in one case and not in the other?

In my opinion, there should be an overlap in both cases

https://user-images.githubusercontent.com/44137912/108828117-bd030c00-75c6-11eb-8592-0b223a711b4a.mp4

**Reply by wnbittle** [reply]
Yes, stacking heavy objects on lighter ones will have this effect and it's expected behavior - don't do it.  If you need stacking, a general rule is to keep all bodies close to the same mass.  Tall stacks of objects _with the same mass_ can also have this effect.  You can reduce the effect by increasing the number of velocity iterations the contact solver uses (by default it's `10` and clearly more iterations means more CPU time):

```java
world.getSettings().setVelocityConstraintSolverIterations(20);
```

> In my opinion, there should be an overlap in both cases

There _is_ overlap in both cases.  The _amount_ of overlap is different.  If you are curious about the _why_, this [video from Erin Catto at GDC](https://www.youtube.com/watch?v=SHinxAhv1ZE) does a great job of explaining this behavior.
